packages/loot-core/src/server/budget/util.js

```diff
@@ -20,3 +20,23 @@ export function unflatten2(arr) {
   return res;
 }
 
+// Note that we don't restrict values to `Number.MIN_SAFE_INTEGER <= value <= Number.MAX_SAFE_INTEGER`
+// where `Number.MAX_SAFE_INTEGER == 2^53 - 1` but a smaller range over `-(2^43-1) <= value <= 2^43 - 1`.
+// This ensure that the number is accurate not just for the integer component but for 3 decimal places also.
+//
+// This gives us the guarantee that can use `safeNumber` on number whether they are unscaled user inputs
+// or they have been converted to integers (using `amountToInteger`).
+
+const MAX_SAFE_NUMBER = 2**43 - 1
+const MIN_SAFE_NUMBER = -MAX_SAFE_NUMBER
+
+export function safeNumber(value) {
```

Comment:
I was thinking of `safeNumber` only checking the range and not using `number`. we'll want to use this outside the budget as well for transactions and such. it should probably live in `shared`